### ISSUE TITLE
[ONNXRuntimeError] : 2 : INVALID_ARGUMENT : Got invalid dimensions for input

### ISSUE THREAD

**Issue by 0x4E69676874466F78** [joeyballentine/chaiNNer#808]
```
An error occurred in a ONNX Upscale Image node:

[ONNXRuntimeError] : 2 : INVALID_ARGUMENT : Got invalid dimensions for input: input for the following indices
 index: 1 Got: 1 Expected: 3
 Please fix either the inputs or the model.

Input values (partial):
• Image: RGB image 2176x2176
• Tile Size Target: 0
```
Models: https://github.com/armory3d/armorlab_models/releases/tag/2022.8
Simple project: [armorlabtest.zip](https://github.com/joeyballentine/chaiNNer/files/9396734/armorlabtest.zip)

**Comment by 0x4E69676874466F78** [comment]
I've tried other Tile Size Targets, no effect. It lacks some more parameters.

**Comment by joeyballentine** [comment]
We don't support things that take in parameters. I don't even know what this model is or what it's for. The only kinds of models officially supported are super resolution models.

**Comment by 0x4E69676874466F78** [comment]
@joeyballentine these models are for generating PBR textures (basecolor, height, normal, occlusion, roughness).
Where in code could I set these parameters myself?

**Comment by 0x4E69676874466F78** [comment]
https://github.com/armory3d/armorlab/blob/a3fbea595478637de59efac7f7f38c746e57a08f/Sources/arm/node/brush/PhotoToPBRNode.hx#L6-L90

**Comment by joeyballentine** [comment]
It would be where we actually run the ONNX model. Though I don't know how to actually set the other parameters as I have not tried to do so before.

**Comment by 0x4E69676874466F78** [comment]
As far as I understand this is just the output size.

**Comment by joeyballentine** [comment]
It's missing two parameters

**Comment by 0x4E69676874466F78** [comment]
Yes, output size (width, height, default: 2048, 2048).
https://github.com/armory3d/armorlab/blob/a3fbea595478637de59efac7f7f38c746e57a08f/Sources/arm/node/brush/PhotoToPBRNode.hx#L78

**Comment by joeyballentine** [comment]
kha.Image.fromBytes is not where it actually runs the model

**Comment by 0x4E69676874466F78** [comment]
There is a very confusing code scattered over many subprojects.
> There is a very confusing code scattered over many subprojects.
Yes.
https://github.com/armory3d/armorlab/blob/a3fbea595478637de59efac7f7f38c746e57a08f/Sources/arm/node/brush/PhotoToPBRNode.hx#L64
https://github.com/armory3d/armorcore/blob/d3221fa81864ba03f0648084cafe3af39ece9a5b/Sources/main.cpp#L3595
https://github.com/armory3d/armorcore/blob/d3221fa81864ba03f0648084cafe3af39ece9a5b/Sources/main.cpp#L3103-L3127
...




https://github.com/armory3d/kha_tensorflow

**Comment by theflyingzamboni** [comment]
The issue is actually just with how we get the number of input channels. chaiNNer is expecting the models to use dynamic axes, where the shape is ["batch_size", 3 (or whatever int value for the number of channels), "height", "width"], the values in quotes being string literals. So the check to get the shape order (bchw vs. bhwc) currently relies on determining which index (1 or 3) contains the int (which because of the dynamic axes is assumed to be in_nc). Since this model uses static axes, all values are ints, and the code returns 1 (at index 0) as the number of input channels instead of 3. Hardcoding 3 allowed this model to run in my testing.

Basically, to get something like this to work, we need another way to get in_nc that doesn't assume dynamic axes.

![image](https://user-images.githubusercontent.com/55669985/186008117-e9a4f0de-497f-486b-b83d-a6c1a29c5c13.png)

**Comment by joeyballentine** [comment]
Oh, that's interesting. I assumed it was like the topaz models that took in other parameters (which we can't support currently) 

So I guess we just need a way to check if it has static axis instead of always assuming dynamic? Do you think that's possible?

**Comment by theflyingzamboni** [comment]
Like I edited in after you probably read that, we may just be able to check whether index 1 is an int <= 4, because I *think* it's unlikely that we will get a model with a static height of 4.

**Comment by joeyballentine** [comment]
I see. Batch would always come first I assume. bhwc and hchw are the two standard ones I know of. Anything else would be nonstandard

**Comment by joeyballentine** [comment]
So I think your idea would work. I don't think any model would be for that small of an image size

**Comment by 0x4E69676874466F78** [comment]
@theflyingzamboni thanks!

**Comment by theflyingzamboni** [comment]
@joeyballentine If you can't think of an issue, I can go ahead and PR that change.

(Side node, really weird that this model hard-codes a size of 2176x2176. What uses that dims, and why so limited?)

**Comment by 0x4E69676874466F78** [comment]
@theflyingzamboni I just know that in Armorlab you can choose 2K or 4K or 8K or 16K

![изображение](https://user-images.githubusercontent.com/4449851/186019109-2258414c-2b92-43ff-b748-f769223d6bd1.png)

**Comment by joeyballentine** [comment]
> What uses those dims, and why so limited?
It might be trained for that resolution specifically? Idk. maybe the arch is such that it needs to have that dimension

**Comment by 0x4E69676874466F78** [comment]
@theflyingzamboni 
> Hardcoding 3 allowed this model to run in my testing.

```
        return convenient_upscale(
            img,
            in_nc,
            lambda i: self.upscale(i, session, split_factor, change_shape),
        )
```
in_nc -> 3?

If so, then the model starts but does not work correctly.

**Comment by 0x4E69676874466F78** [comment]
basecolor from armorlab:
![изображение](https://user-images.githubusercontent.com/4449851/186022668-ab2cb3cf-0dcd-41ad-adb6-daaa977eb169.png)

**Comment by theflyingzamboni** [comment]
> @theflyingzamboni I just know that in Armorlab you can choose 2K or 4K or 8K or 16K
2k is 2048 though. 2176  just seems like such an unusual number to target. Not that it really matters, except you can't pick a tile size, since the axes aren't dynamic.

> basecolor from armorlab
Hmm, are you sure something isn't going wrong there? Those are definitely not the base colors of that texture. I also ran all the other models, and they look basically like I would expect them to, I think (though I'm not positive, the normal is a much darker blue than I'm used to).

**Comment by theflyingzamboni** [comment]
How did you get access to this? Their download page just says "In construction".

**Comment by 0x4E69676874466F78** [comment]
@theflyingzamboni 
> Hmm, are you sure something isn't going wrong there? 

Yes. The base color should not have shadows and highlights, it is a "flat" colors. Normals... wait sec...

> How did you get access to this? 

It needs to be built from source, or I can upload the finished build.

**Comment by 0x4E69676874466F78** [comment]
@theflyingzamboni 

![изображение](https://user-images.githubusercontent.com/4449851/186028919-5c851cf6-2102-4137-8351-6f9e538b4b75.png)

![изображение](https://user-images.githubusercontent.com/4449851/186028889-fba4b4c7-60c6-4a0b-813a-cbdafdf7569d.png)

**Comment by theflyingzamboni** [comment]
Yeah, that looks like more what I'd expect from a normal. If it requires more than just us getting correct dimensions, this may or may not be something we can reasonably implement. Will need to look into it more.

**Comment by 0x4E69676874466F78** [comment]
Maybe it works in a linear color space?

**Comment by 0x4E69676874466F78** [comment]
[ArmorLabRelease.zip](https://github.com/joeyballentine/chaiNNer/files/9398007/ArmorLabRelease.zip) (7zip LZMA2)
drop models in data/models folder

**Comment by theflyingzamboni** [comment]
https://github.com/armory3d/armorlab/blob/a3fbea595478637de59efac7f7f38c746e57a08f/Sources/arm/node/brush/PhotoToPBRNode.hx#L63-L76

I'm reasonably sure that this is not going to be possible to use in chaiNNer. From what I can tell here, it seems the program does additional data transformations outside of just using the model.

**Comment by 0x4E69676874466F78** [comment]
And yes, I was careless, sorry.
But I'll ask the author what kind of magic he's doing there.

**Comment by joeyballentine** [comment]
Closing this for now as these models seem outside chaiNNer's scope

**Comment by 0x4E69676874466F78** [comment]
![изображение](https://user-images.githubusercontent.com/4449851/186465206-3a30a372-5ca2-4361-b311-b1c3f999fddf.png)
![изображение](https://user-images.githubusercontent.com/4449851/186465284-347c099c-babd-4dfa-a84b-b687da7e4810.png)

**Comment by joeyballentine** [comment]
Yeah they definitely are doing stuff other than just purely running the model

**Comment by 0x4E69676874466F78** [comment]
Every pixel (pixel_channel / 255 - 0.5) / 0.5 = result "linear space".
if pixel_channel = 0 then result -1
if pixel_channel = 128 then result 0,003921...
if pixel_channel = 255 then result 1
and then it does the reverse transformation (pixel_channel * 0.5 + 0.5) * 255).

@joeyballentine, Is there a way to iterate over each pixel in this way?

**Comment by 0x4E69676874466F78** [comment]
How do you store a color image in f32?

**Comment by joeyballentine** [comment]
You can't iterate over the pixels with just chainner nodes alone. This is fairly simple with numpy array operations though.
